src/creatures/monsters/monster.cpp

```diff
@@ -64,6 +64,37 @@ void Monster::removeList() {
 	g_game().removeMonster(static_self_cast<Monster>());
 }
 
+const std::string &Monster::getName() const {
+	if (name.empty()) {
+		return mType->name;
+	}
+	return name;
+}
+
+void Monster::setName(const std::string &name) {
+	if (getName() == name) {
+		return;
+	}
+
+	this->name = name;
+
+	// NOTE: Due to how client caches known creatures,
+	// it is not feasible to send creature update to everyone that has ever met it
+	auto spectators = Spectators().find<Player>(position, true);
+	for (const auto spectator : spectators) {
```

Comment:
Use reference to avoid unnecessary copying.
```suggestion
	for (const auto &spectator : spectators) {
```